worlds/yachtdice/Rules.py

```diff
@@ -0,0 +1,253 @@
+        self.state = state
+
+    def count(self, item: str, player: Optional[str] = None) -> int:
+        return self.state.count(item)
+
+
+def extract_progression(state, player, options):
+    # method to obtain a list of what items the player has.
+    # this includes categories, dice, rolls and score multiplier etc.
+    # First, we convert the state if it's a list, so we can use state.count(item, player)
+    if isinstance(state, list):
+        state = ListState(state=state)
+
+    number_of_dice = (
+        state.count("Dice", player)
+        + state.count("Dice Fragment", player) // options.number_of_dice_fragments_per_dice.value
+    )
+    number_of_rerolls = (
+        state.count("Roll", player)
+        + state.count("Roll Fragment", player) // options.number_of_roll_fragments_per_roll.value
+    )
+    number_of_fixed_mults = state.count("Fixed Score Multiplier", player)
+    number_of_step_mults = state.count("Step Score Multiplier", player)
+    categories = []
+    for category_name, category_value in category_mappings.items():
+        if state.count(category_name, player) >= 1:
+            categories += [Category(category_value, state.count(category_name, player))]
+    extra_points_in_logic = state.count("1 Point", player)
+    extra_points_in_logic += state.count("10 Points", player) * 10
+    extra_points_in_logic += state.count("100 Points", player) * 100
+
+    return [
+        categories,
+        number_of_dice,
+        number_of_rerolls,
+        number_of_fixed_mults * 0.1,
+        number_of_step_mults * 0.01,
+        extra_points_in_logic,
+    ]
+
+
+# We will store the results of this function as it is called often for the same parameters.
+
+
+yachtdice_cache = {}
+
+
+# Function that returns the feasible score in logic based on items obtained.
+def dice_simulation_strings(categories, num_dice, num_rolls, fixed_mult, step_mult, diff):
+    tup = tuple(
+        [
+            tuple(sorted([c.name + str(c.quantity) for c in categories])),
+            num_dice,
+            num_rolls,
+            fixed_mult,
+            step_mult,
+            diff,
+        ]
+    )  # identifier
+
+    # if already computed, return the result
+    if tup in yachtdice_cache.keys():
+        return yachtdice_cache[tup]
+
+    # sort categories because for the step multiplier, you will want low-scoring categories first
+    categories.sort(key=lambda category: category.mean_score(num_dice, num_rolls))
+
+    # function to add two discrete distribution.
+    # defaultdict is a dict where you don't need to check if an id is present, you can just use += (lot faster)
+    def add_distributions(dist1, dist2):
+        combined_dist = defaultdict(float)
+        for val1, prob1 in dist1.items():
+            for val2, prob2 in dist2.items():
+                combined_dist[val1 + val2] += prob1 * prob2
+        return dict(combined_dist)
+
+    # function to take the maximum of "times" i.i.d. dist1.
```

Comment:
### Welcome to my territory! Glad you're asking :)
i.i.d. means independent(ly) and identical(ly) distributed.

So let's say you roll for Tiny Straight and you have a 30% chance to get it (20 points).
Now you could write this as a "random variable", say $X$, which is 0 with probability 0.7 and 20 with probability 0.3. $X$ is a (probability) distribution.

Now what if you would get two shots at the Tiny Straight? Mathematically, you could define two **i.i.d. random variables** of $X$, let's call them $X_1$ and $X_2$. 
If we get two shots, we are interested in $\max(X_1, X_2)$, which is also a random variable, in this case:
$\max(X_1, X_2) = 0$ with probability $0.7 \cdot 0.7$
$\max(X_1, X_2) = 20$ with probability $1 - 0.7 \cdot 0.7$
Although you would probably define $Z = \max(X_1, X_2)$ and say $\mathbb{P}[Z = 20] = 1- {0.7}^2$ and 0 otherwise.
 
Later on in the logic we also sum random variables. For example $X$ and $Y$. If $X$ is $i$ with probability $p_i$ and $Y$ is $j$ with probability $q_j$, $$\mathbb{P}[X+Y=k] = \sum_i  p_i \cdot q_{k-i}$$. This sums all possibilities that $X$ could be ($i$), and then puts $Y$ to be $k-i$, so that X+Y equals $i+k-i = k$.

I'm abusing notation a little bit but I'm sure the maths Gods still appreciate my effort.

Now what was I doing........

Reply:
HELP

Reply:
WHERE IS THE RESOLVE BUTTON AAAAAAAAAAAAAAAAA